File: apps/web/src/lib/api/stats.ts
/** Types and fetchers for the /stats resource — mirrors app/schemas/stats.py exactly. */
import { http } from './http'
import type { Surface } from './matches'

export interface StatsFilterParams {
  date_from?: string
  date_to?: string
  surface?: Surface
  opponent_id?: number
  club_id?: number
  tournament_id?: number
  [key: string]: string | number | undefined
}

export interface WinRateStat {
  matches: number
  wins: number
  losses: number
  win_rate: number | null
}

export interface SurfaceWinRate extends WinRateStat {
  surface: Surface
}

export type PeriodGranularity = 'month' | 'season'

export interface PeriodWinRate extends WinRateStat {
  period: string
}

export interface StreakStats {
  current_streak_type: 'Win' | 'Loss' | null
  current_streak_length: number
  longest_win_streak: number
  longest_loss_streak: number
}

export interface RecordStat {
  played: number
  wins: number
  losses: number
  win_rate: number | null
}

export function getWinRate(params: StatsFilterParams = {}, signal?: AbortSignal) {
  return http.get<WinRateStat>('/stats/win-rate', { params, signal })
}

export function getWinRateBySurface(params: StatsFilterParams = {}, signal?: AbortSignal) {
  return http.get<SurfaceWinRate[]>('/stats/win-rate/by-surface', { params, signal })
}

export function getWinRateByPeriod(
  granularity: PeriodGranularity = 'month',
  params: StatsFilterParams = {},
  signal?: AbortSignal,
) {
  return http.get<PeriodWinRate[]>('/stats/win-rate/by-period', {
    params: { ...params, granularity },
    signal,
  })
}

export function getStreaks(params: StatsFilterParams = {}, signal?: AbortSignal) {
  return http.get<StreakStats>('/stats/streaks', { params, signal })
}

export function getTiebreaks(params: StatsFilterParams = {}, signal?: AbortSignal) {
  return http.get<RecordStat>('/stats/tiebreaks', { params, signal })
}

File: apps/api/src/app/schemas/stats.py
"""Pydantic response schemas for the /stats endpoint family.

Every shape here is derived on read from matches/sets (see app/stats.py) —
nothing under /stats is stored.
"""

from typing import Literal

from pydantic import BaseModel

from app.models.enums import Surface


class WinRateStat(BaseModel):
    matches: int
    wins: int
    losses: int
    win_rate: float | None


class SurfaceWinRate(WinRateStat):
    surface: Surface


class OpponentWinRate(WinRateStat):
    opponent_id: int
    opponent_name: str


class ClubWinRate(WinRateStat):
    club_id: int
    club_name: str


class TournamentWinRate(WinRateStat):
    tournament_id: int
    tournament_name: str


class PeriodWinRate(WinRateStat):
    # "YYYY-MM" for month granularity, the tournament's season string for season granularity.
    period: str


class StreakStats(BaseModel):
    current_streak_type: Literal["Win", "Loss"] | None
    current_streak_length: int
    longest_win_streak: int
    longest_loss_streak: int


class RecordStat(BaseModel):
    """A win/loss record for a subset of played sets or matches."""

    played: int
    wins: int
    losses: int
    win_rate: float | None


class GamesRatio(BaseModel):
    games_won: int
    games_lost: int
    ratio: float | None


class MatchesPerMonth(BaseModel):
    period: str
    matches: int
